horovod/run/run.py

```diff
@@ -0,0 +1,425 @@
+# Copyright 2019 Uber Technologies, Inc. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import argparse
+import os
+import sys
+
+import six
+
+try:
+    from shlex import quote
+except ImportError:
+    from pipes import quote
+
+import horovod
+from horovod.run.common.util import codec, safe_shell_exec, timeout, secret
+from horovod.run.driver import driver_service
+from horovod.run.task import task_service
+from horovod.run.util import cache, threads
+
+# Cached information of horovodrun functions be stored in this directory
+CACHE_FOLDER = os.path.join(os.path.expanduser('~'), '.horovod')
+
+# Cache entries will be stale if they are older than this number of minutes
+CACHE_STALENESS_THRESHOLD_MINUTES = 60
+
+# Maximum number of concurrent worker threads to prevent from over saturation.
+MAX_CONCURRENT_EXECUTIONS = 20
+
+# Number of retries for sshing into the hosts
+SSH_RETRIES = 5
+
+
+def use_cache():
+    """
+    If decorates a function, and if cache_disabled is set, it will store the
```

Comment:
Should it say "if fn_cache is set"?

Reply:
Also, should we move this to `cache.py`?

Reply:
done